File: ruby/lib/remitmd/mock.rb
# frozen_string_literal: true

require "securerandom"
require "time"
require "bigdecimal"
require "bigdecimal/util"

module Remitmd
  # In-memory test double for remit.md. Zero network, zero latency, deterministic.
  #
  # MockRemit gives you a RemitWallet backed by in-memory state so you can test
  # your agent's payment logic without a live API or spending real USDC.
  #
  # @example
  #   mock   = Remitmd::MockRemit.new
  #   wallet = mock.wallet
  #
  #   wallet.pay("0x0000000000000000000000000000000000000001", 1.50)
  #
  #   mock.was_paid?("0x0000000000000000000000000000000000000001", 1.50) # => true
  #   mock.total_paid_to("0x0000000000000000000000000000000000000001")   # => BigDecimal("1.5")
  #   mock.transaction_count                                              # => 1
  class MockRemit
    MOCK_ADDRESS = "0xMockWallet0000000000000000000000000001"
    DEFAULT_BALANCE = BigDecimal("10000")

    def initialize(balance: DEFAULT_BALANCE)
      @mutex    = Mutex.new
      @state    = initial_state(BigDecimal(balance.to_s))
    end

    # Return a RemitWallet backed by this mock. No private key required.
    def wallet
      RemitWallet.new(signer: MockSigner.new(MOCK_ADDRESS), transport: MockTransport.new(@state, @mutex))
    end

    # Reset all state. Call between test cases to prevent state leakage.
    def reset
      @mutex.synchronize { @state.replace(initial_state(DEFAULT_BALANCE)) }
    end

    # Override the simulated USDC balance.
    def set_balance(amount)
      @mutex.synchronize { @state[:balance] = BigDecimal(amount.to_s) }
    end

    # Current mock balance.
    def balance
      @mutex.synchronize { @state[:balance] }
    end

    # All transactions recorded.
    def transactions
      @mutex.synchronize { @state[:transactions].dup }
    end

    # Total number of transactions recorded.
    def transaction_count
      @mutex.synchronize { @state[:transactions].length }
    end

    # True if a payment of exactly +amount+ USDC was sent to +recipient+.
    def was_paid?(recipient, amount)
      d = BigDecimal(amount.to_s)
      @mutex.synchronize do
        @state[:transactions].any? do |tx|
          tx.to.casecmp(recipient).zero? && tx.amount == d
        end
      end
    end

    # Sum of all USDC sent to +recipient+.
    def total_paid_to(recipient)
      @mutex.synchronize do
        @state[:transactions]
          .select { |tx| tx.to.casecmp(recipient).zero? }
          .sum(BigDecimal("0"), &:amount)
      end
    end

    private

    def initial_state(balance)
      {
        balance:           balance,
        transactions:      [],
        escrows:           {},
        tabs:              {},
        streams:           {},
        bounties:          {},
        deposits:          {},
        pending_invoices:  {},
      }
    end
  end

  # ─── Internal: MockSigner ──────────────────────────────────────────────────

  class MockSigner
    include Signer

    def initialize(addr)
      @address = addr
    end

    attr_reader :address

    def sign(_message)
      "0x" + SecureRandom.hex(32) + SecureRandom.hex(32) + "1b"
    end
  end

  # ─── Internal: MockTransport ───────────────────────────────────────────────

  class MockTransport
    def initialize(state, mutex)
      @state = state
      @mutex = mutex
    end

    def get(path)
      dispatch("GET", path, nil)
    end

    def post(path, body = nil)
      dispatch("POST", path, body)
    end

    private

    def dispatch(method, path, body)
      @mutex.synchronize { handle(method, path, body || {}) }
    end

    def handle(method, path, b) # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/MethodLength, Metrics/PerceivedComplexity
      case [method, path]

      # Contracts (for auto_permit)
      in ["GET", "/contracts"]
        {
          "chain_id"       => ChainId::BASE_SEPOLIA,
          "usdc"           => "0xMockUSDC0000000000000000000000000000001",
          "router"         => "0xMockRouter000000000000000000000000001",
          "escrow"         => "0xMockEscrow000000000000000000000000001",
          "tab"            => "0xMockTab00000000000000000000000000001",
          "stream"         => "0xMockStream000000000000000000000000001",
          "bounty"         => "0xMockBounty000000000000000000000000001",
          "deposit"        => "0xMockDeposit00000000000000000000000001",
          "fee_calculator" => "0xMockFeeCalc0000000000000000000000001",
          "key_registry"   => "0xMockKeyReg000000000000000000000000001",
        }

      # Balance
      in ["GET", "/wallet/balance"]
        {
          "usdc"       => @state[:balance].to_s("F"),
          "address"    => MockRemit::MOCK_ADDRESS,
          "chain_id"   => ChainId::BASE_SEPOLIA,
          "updated_at" => now,
        }

      # Direct payment
      in ["POST", "/payments/direct"]
        to     = fetch!(b, :to)
        amount = decimal!(b, :amount)
        check_balance!(amount)
        @state[:balance] -= amount
        tx = make_tx(to: to, amount: amount, memo: b[:memo].to_s)
        @state[:transactions] << tx
        tx_hash(tx)

      # Invoice create (step 1 of escrow)
      in ["POST", "/invoices"]
        id = fetch!(b, :id)
        @state[:pending_invoices][id] = b
        { "id" => id, "status" => "pending" }

      # Escrow create (step 2 — fund with invoice_id)
      in ["POST", "/escrows"]
        invoice_id = fetch!(b, :invoice_id)
        inv = @state[:pending_invoices].delete(invoice_id)
        raise not_found(RemitError::ESCROW_NOT_FOUND, invoice_id) unless inv
        payee  = (inv[:to_agent] || inv["to_agent"]).to_s
        amount = decimal!(inv, :amount)
        memo   = (inv[:task] || inv["task"]).to_s
        check_balance!(amount)
        @state[:balance] -= amount
        esc = make_escrow(payee: payee, amount: amount, memo: memo, id: invoice_id)
        @state[:escrows][esc.id] = esc
        escrow_hash(esc)

      # Escrow claim-start
      in ["POST", path] if path.end_with?("/claim-start") && path.include?("/escrows/")
        id  = extract_id(path, "/escrows/", "/claim-start")
        esc = @state[:escrows].fetch(id) { raise not_found(RemitError::ESCROW_NOT_FOUND, id) }
        escrow_hash(esc)

      # Escrow release
      in ["POST", path] if path.end_with?("/release") && path.include?("/escrows/")
        id  = extract_id(path, "/escrows/", "/release")
        esc = @state[:escrows].fetch(id) { raise not_found(RemitError::ESCROW_NOT_FOUND, id) }
        new_esc = update_escrow(esc, status: EscrowStatus::RELEASED)
        @state[:escrows][id] = new_esc
        tx = make_tx(from: esc.payer, to: esc.payee, amount: esc.amount)
        @state[:transactions] << tx
        tx_hash(tx)

      # Escrow cancel
      in ["POST", path] if path.end_with?("/cancel") && path.include?("/escrows/")
        id  = extract_id(path, "/escrows/", "/cancel")
        esc = @state[:escrows].fetch(id) { raise not_found(RemitError::ESCROW_NOT_FOUND, id) }
        new_esc = update_escrow(esc, status: EscrowStatus::CANCELLED)
        @state[:escrows][id] = new_esc
        @state[:balance] += esc.amount
        tx = make_tx(from: esc.payer, to: esc.payer, amount: esc.amount, memo: "escrow cancelled")
        @state[:transactions] << tx
        tx_hash(tx)

      # Escrow get
      in ["GET", path] if path.start_with?("/escrows/")
        id  = path.delete_prefix("/escrows/")
        esc = @state[:escrows].fetch(id) { raise not_found(RemitError::ESCROW_NOT_FOUND, id) }
        escrow_hash(esc)

      # Tab create
      in ["POST", "/tabs"]
        provider    = fetch!(b, :provider)
        limit       = decimal!(b, :limit_amount)
        tab = make_tab(provider: provider, limit: limit)
        @state[:tabs][tab.id] = tab
        tab_hash(tab)

      # Tab debit (legacy off-chain)
      in ["POST", path] if path.end_with?("/debit") && path.include?("/tabs/")
        id     = extract_id(path, "/tabs/", "/debit")
        amount = decimal!(b, :amount)
        tab    = @state[:tabs].fetch(id) { raise not_found(RemitError::TAB_NOT_FOUND, id) }
        new_tab = update_tab(tab, used: tab.used + amount, remaining: tab.remaining - amount)
        @state[:tabs][id] = new_tab
        {
          "tab_id"     => id,
          "amount"     => amount.to_s("F"),
          "cumulative" => new_tab.used.to_s("F"),
          "call_count" => 0,
          "memo"       => b[:memo].to_s,
          "sequence"   => 1,
          "signature"  => "0x00",
        }

      # Tab charge (EIP-712 signed)
      in ["POST", path] if path.end_with?("/charge") && path.include?("/tabs/")
        id     = extract_id(path, "/tabs/", "/charge")
        amount = decimal!(b, :amount)
        tab    = @state[:tabs].fetch(id) { raise not_found(RemitError::TAB_NOT_FOUND, id) }
        new_tab = update_tab(tab, used: tab.used + amount, remaining: tab.remaining - amount)
        @state[:tabs][id] = new_tab
        {
          "tab_id"     => id,
          "amount"     => amount.to_s("F"),
          "cumulative" => new_tab.used.to_s("F"),
          "call_count" => (b[:call_count] || 1).to_i,
          "memo"       => b[:memo].to_s,
          "sequence"   => 1,
          "signature"  => "0x00",
        }

      # Tab close (settle)
      in ["POST", path] if path.end_with?("/close") && path.include?("/tabs/")
        id  = extract_id(path, "/tabs/", "/close")
        tab = @state[:tabs].fetch(id) { raise not_found(RemitError::TAB_NOT_FOUND, id) }
        new_tab = update_tab(tab, status: TabStatus::SETTLED)
        @state[:tabs][id] = new_tab
        tab_hash(new_tab)

      # Stream create
      in ["POST", "/streams"]
        payee = fetch!(b, :payee)
        rate_per_second = decimal!(b, :rate_per_second)
        max_total = decimal!(b, :max_total)
        check_balance!(max_total)
        @state[:balance] -= max_total
        s = make_stream(recipient: payee, rate_per_sec: rate_per_second, deposited: max_total)
        @state[:streams][s.id] = s
        stream_hash(s)

      # Stream close
      in ["POST", path] if path.end_with?("/close") && path.include?("/streams/")
        id  = extract_id(path, "/streams/", "/close")
        s   = @state[:streams].fetch(id) { raise not_found(RemitError::STREAM_NOT_FOUND, id) }
        stream_hash(s)

      # Stream withdraw
      in ["POST", path] if path.end_with?("/withdraw") && path.include?("/streams/")
        id  = extract_id(path, "/streams/", "/withdraw")
        s   = @state[:streams].fetch(id) { raise not_found(RemitError::STREAM_NOT_FOUND, id) }
        tx  = make_tx(from: s.sender, to: s.recipient, amount: s.deposited, memo: "stream withdraw")
        @state[:transactions] << tx
        tx_hash(tx)

      # Bounty create
      in ["POST", "/bounties"]
        amount           = decimal!(b, :amount)
        task_description = fetch!(b, :task_description)
        check_balance!(amount)
        @state[:balance] -= amount
        bnt = make_bounty(amount: amount, task_description: task_description)
        @state[:bounties][bnt.id] = bnt
        bounty_hash(bnt)

      # Bounty submit
      in ["POST", path] if path.end_with?("/submit") && path.include?("/bounties/")
        id = extract_id(path, "/bounties/", "/submit")
        bnt = @state[:bounties].fetch(id) { raise not_found(RemitError::BOUNTY_NOT_FOUND, id) }
        {
          "id"            => 1,
          "bounty_id"     => id,
          "submitter"     => MockRemit::MOCK_ADDRESS,
          "evidence_hash" => (b[:evidence_hash] || "0x00").to_s,
          "status"        => "pending",
          "created_at"    => now,
        }

      # Bounty award
      in ["POST", path] if path.end_with?("/award") && path.include?("/bounties/")
        id            = extract_id(path, "/bounties/", "/award")
        submission_id = (b[:submission_id] || b["submission_id"]).to_i
        bnt           = @state[:bounties].fetch(id) { raise not_found(RemitError::BOUNTY_NOT_FOUND, id) }
        new_bnt = update_bounty(bnt, status: BountyStatus::AWARDED)
        @state[:bounties][id] = new_bnt
        bounty_hash(new_bnt)

      # Deposit create
      in ["POST", "/deposits"]
        provider = fetch!(b, :provider)
        amount   = decimal!(b, :amount)
        check_balance!(amount)
        @state[:balance] -= amount
        dep = make_deposit(provider: provider, amount: amount)
        @state[:deposits][dep.id] = dep
        deposit_hash(dep)

      # Deposit return
      in ["POST", path] if path.end_with?("/return") && path.include?("/deposits/")
        id  = extract_id(path, "/deposits/", "/return")
        dep = @state[:deposits].fetch(id) { raise not_found(RemitError::DEPOSIT_NOT_FOUND, id) }
        @state[:balance] += dep.amount
        tx = make_tx(from: dep.provider, to: dep.depositor, amount: dep.amount, memo: "deposit returned")
        @state[:transactions] << tx
        tx_hash(tx)

      # Reputation
      in ["GET", path] if path.start_with?("/reputation/")
        addr = path.delete_prefix("/reputation/")
        {
          "address"           => addr,
          "score"             => 750,
          "total_paid"        => "1000.0",
          "total_received"    => "500.0",
          "transaction_count" => 42,
          "member_since"      => now,
        }

      # Spending summary
      in ["GET", "/wallet/spending"]
        total = @state[:transactions].sum(BigDecimal("0"), &:amount)
        count = @state[:transactions].length
        {
          "address"        => MockRemit::MOCK_ADDRESS,
          "period"         => "month",
          "total_spent"    => total.to_s("F"),
          "total_fees"     => (BigDecimal("0.001") * count).to_s("F"),
          "tx_count"       => count,
          "top_recipients" => [],
        }

      # Budget
      in ["GET", "/wallet/budget"]
        {
          "daily_limit"       => "10000.0",
          "daily_used"        => "0.0",
          "daily_remaining"   => "10000.0",
          "monthly_limit"     => "100000.0",
          "monthly_used"      => "0.0",
          "monthly_remaining" => "100000.0",
          "per_tx_limit"      => "1000.0",
        }

      # History
      in ["GET", path] if path.start_with?("/wallet/history")
        txs = @state[:transactions]
        {
          "items"    => txs.map { |tx| tx_hash(tx) },
          "total"    => txs.length,
          "page"     => 1,
          "per_page" => 50,
          "has_more" => false,
        }

      # Intents
      in ["POST", "/intents"]
        to     = fetch!(b, :to)
        amount = decimal!(b, :amount)
        {
          "id"         => new_id("int"),
          "from"       => MockRemit::MOCK_ADDRESS,
          "to"         => to,
          "amount"     => amount.to_s("F"),
          "type"       => b[:type] || "direct",
          "expires_at" => now,
          "created_at" => now,
        }

      else
        {}
      end
    end

    # ─── Builder helpers ────────────────────────────────────────────────────

    def make_tx(to:, amount:, from: MockRemit::MOCK_ADDRESS, memo: "")
      Transaction.new(
        "id"           => new_id("tx"),
        "tx_hash"      => "0x#{SecureRandom.hex(32)}",
        "from"         => from,
        "to"           => to,
        "amount"       => amount.to_s("F"),
        "fee"          => "0.001",
        "memo"         => memo,
        "chain_id"     => ChainId::BASE_SEPOLIA,
        "block_number" => 0,
        "created_at"   => now,
      )
    end

    def make_escrow(payee:, amount:, memo: "", id: nil)
      Escrow.new(
        "id"         => id || new_id("esc"),
        "payer"      => MockRemit::MOCK_ADDRESS,
        "payee"      => payee,
        "amount"     => amount.to_s("F"),
        "fee"        => "0.001",
        "status"     => EscrowStatus::FUNDED,
        "memo"       => memo,
        "milestones" => [],
        "splits"     => [],
        "created_at" => now,
      )
    end

    def make_tab(provider:, limit:)
      Tab.new(
        "id"           => new_id("tab"),
        "opener"       => MockRemit::MOCK_ADDRESS,
        "provider"     => provider,
        "limit_amount" => limit.to_s("F"),
        "used"         => "0",
        "remaining"    => limit.to_s("F"),
        "status"       => TabStatus::OPEN,
        "created_at"   => now,
      )
    end

    def make_stream(recipient:, rate_per_sec:, deposited:)
      Stream.new(
        "id"           => new_id("str"),
        "sender"       => MockRemit::MOCK_ADDRESS,
        "recipient"    => recipient,
        "rate_per_sec" => rate_per_sec.to_s("F"),
        "deposited"    => deposited.to_s("F"),
        "withdrawn"    => "0",
        "status"       => StreamStatus::ACTIVE,
        "started_at"   => now,
      )
    end

    def make_bounty(amount:, task_description:)
      Bounty.new(
        "id"               => new_id("bnt"),
        "poster"           => MockRemit::MOCK_ADDRESS,
        "amount"           => amount.to_s("F"),
        "task_description" => task_description,
        "status"           => BountyStatus::OPEN,
        "winner"           => "",
        "created_at"       => now,
      )
    end

    def make_deposit(provider:, amount:)
      Deposit.new(
        "id"          => new_id("dep"),
        "depositor"   => MockRemit::MOCK_ADDRESS,
        "provider"    => provider,
        "amount"      => amount.to_s("F"),
        "status"      => DepositStatus::LOCKED,
        "created_at"  => now,
      )
    end

    def update_escrow(esc, **changes)
      h = escrow_hash(esc).merge(changes.transform_keys(&:to_s))
      Escrow.new(h)
    end

    def update_tab(tab, **changes)
      h = tab_hash(tab).merge(changes.transform_keys(&:to_s))
      Tab.new(h)
    end

    def update_bounty(bnt, **changes)
      h = bounty_hash(bnt).merge(changes.transform_keys(&:to_s))
      Bounty.new(h)
    end

    def tx_hash(tx)
      {
        "id"           => tx.id,
        "tx_hash"      => tx.tx_hash,
        "from"         => tx.from,
        "to"           => tx.to,
        "amount"       => tx.amount.to_s("F"),
        "fee"          => tx.fee.to_s("F"),
        "memo"         => tx.memo,
        "chain_id"     => tx.chain_id,
        "block_number" => tx.block_number,
        "created_at"   => now,
      }
    end

    def escrow_hash(esc)
      {
        "id"         => esc.id,
        "payer"      => esc.payer,
        "payee"      => esc.payee,
        "amount"     => esc.amount.to_s("F"),
        "fee"        => esc.fee.to_s("F"),
        "status"     => esc.status,
        "memo"       => esc.memo,
        "milestones" => esc.milestones,
        "splits"     => esc.splits,
        "expires_at" => esc.expires_at&.iso8601,
        "created_at" => now,
      }
    end

    def tab_hash(tab)
      {
        "id"           => tab.id,
        "opener"       => tab.opener,
        "provider"     => tab.provider,
        "limit_amount" => tab.limit.to_s("F"),
        "used"         => tab.used.to_s("F"),
        "remaining"    => tab.remaining.to_s("F"),
        "status"       => tab.status,
        "created_at"   => now,
      }
    end

    def stream_hash(s)
      {
        "id"           => s.id,
        "sender"       => s.sender,
        "recipient"    => s.recipient,
        "rate_per_sec" => s.rate_per_sec.to_s("F"),
        "deposited"    => s.deposited.to_s("F"),
        "withdrawn"    => s.withdrawn.to_s("F"),
        "status"       => s.status,
        "started_at"   => now,
      }
    end

    def bounty_hash(bnt)
      {
        "id"               => bnt.id,
        "poster"           => bnt.poster,
        "amount"           => bnt.amount.to_s("F"),
        "task_description" => bnt.task_description,
        "status"           => bnt.status,
        "winner"           => bnt.winner,
        "expires_at"       => bnt.expires_at&.iso8601,
        "created_at"       => now,
      }
    end

    def deposit_hash(dep)
      {
        "id"          => dep.id,
        "depositor"   => dep.depositor,
        "provider"    => dep.provider,
        "amount"      => dep.amount.to_s("F"),
        "status"      => dep.status,
        "expires_at"  => dep.expires_at&.iso8601,
        "created_at"  => now,
      }
    end

    # ─── Utilities ──────────────────────────────────────────────────────────

    def new_id(prefix)
      "#{prefix}_#{SecureRandom.hex(4)}"
    end

    def now
      Time.now.utc.iso8601
    end

    def fetch!(h, key)
      val = h[key] || h[key.to_s]
      raise RemitError.new(RemitError::SERVER_ERROR, "Missing field: #{key}") if val.nil?

      val.to_s
    end

    def decimal!(h, key)
      s = fetch!(h, key)
      BigDecimal(s)
    rescue ArgumentError
      raise RemitError.new(RemitError::INVALID_AMOUNT, "Invalid decimal for #{key}: #{s.inspect}")
    end

    def check_balance!(amount)
      bal = @state[:balance]
      return if bal >= amount
      raise RemitError.new(
        RemitError::INSUFFICIENT_FUNDS,
        "Insufficient balance: have #{bal.to_s("F")} USDC, need #{amount.to_s("F")} USDC",
        context: { balance: bal.to_s("F"), amount: amount.to_s("F") }
      )
    end

    def extract_id(path, prefix, suffix)
      path.delete_prefix(prefix).delete_suffix(suffix)
    end

    def not_found(code, id)
      RemitError.new(code, "#{code}: #{id} not found")
    end
  end
end

File: elixir/lib/remit_md/mock.ex
defmodule RemitMd.MockRemit do
  @moduledoc """
  In-memory mock of the remit.md API for testing. Implements the full
  payment API without network or blockchain - completes in microseconds.

  `MockRemit` is an OTP `GenServer`. Start it under a supervisor or with
  `start_link/1` in a test. Pass the pid to `RemitMd.Wallet.new/1`.

  ## Example

      test "agent pays for a service" do
        {:ok, mock} = RemitMd.MockRemit.start_link()
        RemitMd.MockRemit.set_balance(mock, "0xPayer", "100.00")

        wallet = RemitMd.Wallet.new(mock: mock, address: "0xPayer")
        {:ok, tx} = RemitMd.Wallet.pay(wallet, "0xRecipient", "5.00")

        assert tx.status == "confirmed"
        assert RemitMd.MockRemit.was_paid?(mock, "0xRecipient")
        assert RemitMd.MockRemit.total_paid_to(mock, "0xRecipient") == "5.00"
      end
  """

  use GenServer

  @default_balance "1000.00"
  @default_reputation_score 85
  @fee_rate "0.001" # 0.1%

  # ─── Public API ───────────────────────────────────────────────────────────

  @doc "Start a MockRemit GenServer."
  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, %{}, opts)
  end

  @doc "Set the USDC balance for a wallet address."
  def set_balance(pid, address, amount_usdc) do
    GenServer.call(pid, {:set_balance, address, amount_usdc})
  end

  @doc "Set the reputation score for a wallet address (0–100)."
  def set_reputation(pid, address, score) do
    GenServer.call(pid, {:set_reputation, address, score})
  end

  @doc "Return true if any payment was sent TO the given address."
  def was_paid?(pid, address) do
    GenServer.call(pid, {:was_paid, address})
  end

  @doc "Return the total USDC sent TO the given address across all payments."
  def total_paid_to(pid, address) do
    GenServer.call(pid, {:total_paid_to, address})
  end

  @doc "Return the number of transactions recorded."
  def transaction_count(pid) do
    GenServer.call(pid, :transaction_count)
  end

  @doc "Reset all state (balances, transactions, escrows, etc.)."
  def reset(pid) do
    GenServer.call(pid, :reset)
  end

  # ─── Payment Operations (called by MockTransport) ─────────────────────────

  @doc false
  def do_pay(pid, from, to, amount_usdc, opts) do
    GenServer.call(pid, {:pay, from, to, amount_usdc, opts})
  end

  @doc false
  def do_balance(pid, address) do
    GenServer.call(pid, {:balance, address})
  end

  @doc false
  def do_reputation(pid, address) do
    GenServer.call(pid, {:reputation, address})
  end

  @doc false
  def do_history(pid, address, opts) do
    GenServer.call(pid, {:history, address, opts})
  end

  @doc false
  def do_create_escrow(pid, from, to, amount_usdc, opts) do
    GenServer.call(pid, {:create_escrow, from, to, amount_usdc, opts})
  end

  @doc false
  def do_pay_milestone(pid, from, escrow_id, milestone_id) do
    GenServer.call(pid, {:pay_milestone, from, escrow_id, milestone_id})
  end

  @doc false
  def do_cancel_escrow(pid, from, escrow_id) do
    GenServer.call(pid, {:cancel_escrow, from, escrow_id})
  end

  @doc false
  def do_spending(pid, address, opts) do
    GenServer.call(pid, {:spending, address, opts})
  end

  # ─── GenServer Callbacks ──────────────────────────────────────────────────

  @impl true
  def init(_opts) do
    {:ok, initial_state()}
  end

  @impl true
  def handle_call({:set_balance, address, amount}, _from, state) do
    balances = Map.put(state.balances, address, amount)
    {:reply, :ok, %{state | balances: balances}}
  end

  def handle_call({:set_reputation, address, score}, _from, state) do
    reps = Map.put(state.reputations, address, score)
    {:reply, :ok, %{state | reputations: reps}}
  end

  def handle_call({:was_paid, address}, _from, state) do
    result = state.transactions |> Enum.any?(fn tx -> tx["to"] == address end)
    {:reply, result, state}
  end

  def handle_call({:total_paid_to, address}, _from, state) do
    total =
      state.transactions
      |> Enum.filter(fn tx -> tx["to"] == address end)
      |> Enum.reduce(Decimal.new("0"), fn tx, acc ->
        Decimal.add(acc, Decimal.new(tx["amount_usdc"]))
      end)

    {:reply, Decimal.to_string(total), state}
  end

  def handle_call(:transaction_count, _from, state) do
    {:reply, length(state.transactions), state}
  end

  def handle_call(:reset, _from, _state) do
    {:reply, :ok, initial_state()}
  end

  def handle_call({:pay, from, to, amount_usdc, _opts}, _from, state) do
    with :ok <- check_balance(state, from, amount_usdc),
         :ok <- check_not_same(from, to) do
      fee = compute_fee(amount_usdc)
      tx = make_tx("pay", from, to, amount_usdc, fee)
      state = %{state | transactions: [tx | state.transactions]}
      {:reply, {:ok, tx}, state}
    else
      {:error, _} = err -> {:reply, err, state}
    end
  end

  def handle_call({:balance, address}, _from, state) do
    usdc = Map.get(state.balances, address, @default_balance)

    result = %{
      "address"  => address,
      "usdc"     => usdc,
      "chain_id" => 84532
    }

    {:reply, {:ok, result}, state}
  end

  def handle_call({:reputation, address}, _from, state) do
    score = Map.get(state.reputations, address, @default_reputation_score)
    count = state.transactions |> Enum.count(fn tx -> tx["to"] == address || tx["from"] == address end)

    result = %{
      "address"             => address,
      "score"               => score,
      "total_volume_usdc"   => "0.00",
      "successful_txns"     => count,
      "avg_settlement_secs" => 1,
      "member_since"        => "2026-01-01T00:00:00Z"
    }

    {:reply, {:ok, result}, state}
  end

  def handle_call({:history, address, opts}, _from, state) do
    limit  = Keyword.get(opts, :limit, 50)
    offset = Keyword.get(opts, :offset, 0)

    relevant =
      state.transactions
      |> Enum.filter(fn tx -> tx["from"] == address || tx["to"] == address end)

    items = relevant |> Enum.drop(offset) |> Enum.take(limit)

    result = %{
      "items"  => items,
      "total"  => length(relevant),
      "limit"  => limit,
      "offset" => offset
    }

    {:reply, {:ok, result}, state}
  end

  def handle_call({:create_escrow, from, to, amount_usdc, opts}, _from, state) do
    with :ok <- check_balance(state, from, amount_usdc),
         :ok <- check_not_same(from, to) do
      milestones = Keyword.get(opts, :milestones, ["complete"])
      fee = compute_fee(amount_usdc)
      escrow_id = "esc_" <> generate_id()

      escrow = %{
        "escrow_id"   => escrow_id,
        "from"        => from,
        "to"          => to,
        "amount_usdc" => amount_usdc,
        "fee_usdc"    => fee,
        "status"      => "pending",
        "milestones"  => Enum.map(milestones, fn m -> %{"id" => m, "status" => "pending"} end),
        "created_at"  => DateTime.utc_now() |> DateTime.to_iso8601()
      }

      state = %{state | escrows: Map.put(state.escrows, escrow_id, escrow)}
      {:reply, {:ok, escrow}, state}
    else
      {:error, _} = err -> {:reply, err, state}
    end
  end

  def handle_call({:pay_milestone, from, escrow_id, milestone_id}, _from, state) do
    case Map.get(state.escrows, escrow_id) do
      nil ->
        {:reply, {:error, :not_found}, state}

      %{"from" => owner} = _escrow when owner != from ->
        {:reply, {:error, :forbidden}, state}

      escrow ->
        milestones =
          Enum.map(escrow["milestones"], fn m ->
            if m["id"] == milestone_id, do: Map.put(m, "status", "released"), else: m
          end)

        all_done = Enum.all?(milestones, fn m -> m["status"] == "released" end)
        new_status = if all_done, do: "complete", else: "partial"

        updated = %{escrow | "milestones" => milestones, "status" => new_status}
        state = %{state | escrows: Map.put(state.escrows, escrow_id, updated)}
        {:reply, {:ok, updated}, state}
    end
  end

  def handle_call({:cancel_escrow, from, escrow_id}, _from, state) do
    case Map.get(state.escrows, escrow_id) do
      nil ->
        {:reply, {:error, :not_found}, state}

      %{"from" => owner} = _escrow when owner != from ->
        {:reply, {:error, :forbidden}, state}

      escrow ->
        cancelled = Map.put(escrow, "status", "cancelled")
        state = %{state | escrows: Map.put(state.escrows, escrow_id, cancelled)}
        {:reply, {:ok, cancelled}, state}
    end
  end

  def handle_call({:spending, address, opts}, _from, state) do
    limit = Keyword.get(opts, :limit, 30)

    sent =
      state.transactions
      |> Enum.filter(fn tx -> tx["from"] == address end)

    total =
      Enum.reduce(sent, Decimal.new("0"), fn tx, acc ->
        Decimal.add(acc, Decimal.new(tx["amount_usdc"]))
      end)

    result = %{
      "address"           => address,
      "total_spent_usdc"  => Decimal.to_string(total),
      "transaction_count" => length(sent),
      "top_recipients"    => [],
      "period_start"      => DateTime.utc_now() |> DateTime.add(-limit * 86400, :second) |> DateTime.to_iso8601(),
      "period_end"        => DateTime.utc_now() |> DateTime.to_iso8601()
    }

    {:reply, {:ok, result}, state}
  end

  # ─── Helpers ──────────────────────────────────────────────────────────────

  defp initial_state do
    %{
      balances:     %{},
      reputations:  %{},
      transactions: [],
      escrows:      %{}
    }
  end

  defp check_balance(state, address, amount_usdc) do
    balance = Map.get(state.balances, address, @default_balance)

    if Decimal.compare(Decimal.new(balance), Decimal.new(amount_usdc)) == :lt do
      {:error,
       RemitMd.Error.new(
         RemitMd.Error.insufficient_balance(),
         "Balance #{balance} USDC < #{amount_usdc} USDC required"
       )}
    else
      :ok
    end
  end

  defp check_not_same(from, to) do
    if from == to do
      {:error,
       RemitMd.Error.new(
         RemitMd.Error.self_payment(),
         "Payer and payee cannot be the same address"
       )}
    else
      :ok
    end
  end

  defp make_tx(model, from, to, amount_usdc, fee_usdc) do
    %{
      "tx_id"       => "tx_" <> generate_id(),
      "from"        => from,
      "to"          => to,
      "amount_usdc" => amount_usdc,
      "fee_usdc"    => fee_usdc,
      "model"       => model,
      "status"      => "confirmed",
      "tx_hash"     => "0x" <> generate_id() <> generate_id(),
      "chain_id"    => 84532,
      "created_at"  => DateTime.utc_now() |> DateTime.to_iso8601()
    }
  end

  defp compute_fee(amount_usdc) do
    Decimal.new(amount_usdc)
    |> Decimal.mult(Decimal.new(@fee_rate))
    |> Decimal.to_string()
  end

  defp generate_id do
    :crypto.strong_rand_bytes(8) |> Base.encode16(case: :lower)
  end
end
